Making @automattic/components work in isolation

jameslnewell opened the pull request:
#### Changes proposed in this Pull Request

* renamed and moved `$sans`, `$sans-rtl`, `breakpoint` and `clear-fix` variables/mixins into `@automattic/components` with a prefix of `a8c-`
* updated `@automattic/components` to include any relevant styles defined from the Calypso reset/base styles (mostly just `font-family` related stuff)
* added stories for all of the `@automattic/components` to test them in isolation from the Calypso reset/base styles

**Future TODO**:
* Need to come up with a plan for handling `z-index` (used by a handful of the components) that scales across multiple apps

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Check instances of fonts, breakpoints and clear fix across the app
* Check components in the Storybook visually look the same as the components in `/devdocs`

Commit:
copied the necessary CSS to make button work in isolation

Commit:
added typescript support for storybook

Commit:
move necessary typography variables and clear-fix mixin to @automattic/components

sirreal commented:
This needs a rebase and has some failing e2e tests. Can you look into them?

jameslnewell commented:
Rebased to magically fix :allthethings: 🙃

sirreal commented on packages/components/src/styles/_typography.scss:
```diff
@@ -0,0 +1,5 @@
+
+$a8c-fontFamily-sans: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen-Sans', 'Ubuntu', 'Cantarell',
```
According to [CSS/Sass guidelines](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#general-syntax-and-writing-rules), camelCase should be avoided in variables.

> Use hyphens, not underscores or camelCase, when naming things like IDs, classes, variable names, mixins, placeholders. Good: `.site-title`, Bad: `.siteTitle` or `.site_title`.

```suggestion
$a8c-font-family-sans: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen-Sans', 'Ubuntu', 'Cantarell',
```

jsnajdr replied:
I agree with @sirreal -- mixing two case conventions in one name is unpleasant to look at 🙂

sirreal commented on packages/components/src/styles/_typography.scss:
```diff
@@ -0,0 +1,5 @@
+
+$a8c-fontFamily-sans: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen-Sans', 'Ubuntu', 'Cantarell',
+'Helvetica Neue', sans-serif;
+$a8c-fontFamily-sans-rtl: Tahoma, $a8c-fontFamily-sans;
```
```suggestion
$a8c-font-family-sans-rtl: Tahoma, $a8c-font-family-sans;
```

sirreal commented on packages/components/src/styles/_typography.scss:
```diff
@@ -0,0 +1,5 @@
+
+$a8c-fontFamily-sans: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen-Sans', 'Ubuntu', 'Cantarell',
+'Helvetica Neue', sans-serif;
```
This line break with no indentation doesn't look right. Was that manual or automated?

sirreal commented on packages/components/src/styles/_clear-fix.scss:
```diff
@@ -3,7 +3,7 @@
 // ==========================================================================
 
-@mixin clear-fix {
+@mixin a8c-clearFix {
```
According to [CSS/Sass guidelines](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#general-syntax-and-writing-rules), camelCase should be avoided in variables.

> Use hyphens, not underscores or camelCase, when naming things like IDs, classes, variable names, mixins, placeholders. Good: `.site-title`, Bad: `.siteTitle` or `.site_title`.

```suggestion
@mixin a8c-clear-fix {
```

---

I also noticed when [reviewing the guidelines](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#mixins-vs-extends), it seems the clearfix should be an `@extend`. I suspect that no longer holds true with our tree-shaken modular stylesheets 🙅‍♂ 

> * DO use `@extend` when in doubt, using the `%placeholder` syntax. This will produce leaner output.
> * DON'T use mixins for anything that doesn't accept an argument. This is `@extend` territory.
> * DO read [this article](http://miguelcamba.com/blog/2013/07/11/sass-placeholders-versus-mixins-and-extends/) if you don't understand `@extend`.

@jsnajdr Do you think this documentation section is worth revisiting?

jsnajdr replied:
With modular stylesheets, you should never extend from a "real" class, like `@extend .placeholder`. The extended class is unlikely to be present in the currently compiled CSS module.

`@extend %placeholder` is fine -- it's a usage of a reusable building block that doesn't output any CSS by itself. And the `%placeholder` definition will be in one of the "header" files that the compilation includes at the beginning.

The "leaner output" argument can be illustrated by this example:
```
%placeholder {
  color: grey;
}
.stats-placeholder {
  @extend %placeholder;
  width: 100%;
}
.reader-placeholder {
  @extend %placeholder;
  height: 100%;
}
```
This will output CSS that contains the `%placeholder` properties only once:
```
.stats-placeholder,  .reader-placeholder {
  color: grey;
}
.stats-placeholder {
  width: 100%;
}
.reader-placeholder {
  height: 100%;
}
```
Using a mixin would simply duplicate the properties:
```
.stats-placeholder {
  color: grey;
  width: 100%;
}
.reader-placeholder {
  color: grey;
  height: 100%;
}
```
This was an interesting optimization when the whole Calypso stylesheet was a single compilation unit. Then the styles were optimized globally. Now the deduplication is local to the currently compiled module, and no longer so effective.

> Do you think this documentation section is worth revisiting?

The part you quote still seems 100% true. It's just that the recommendation to prefer extend over mixins is much less important now.

sirreal commented on packages/components/tsconfig.json:
```diff
@@ -21,6 +21,6 @@
 		"forceConsistentCasingInFileNames": true
 	},
-	"files": [ "src/index" ],
+	"include": [ "src" ],
```
Why did this need to change? Storybook?

jameslnewell replied:
Yep, specifically `react-docgen-typescript-loader` which is added by the [preset](https://github.com/storybookjs/presets/blob/master/packages/preset-typescript/index.js#L31).

Was there a reason we were including the single file? If its an issue I can create/extend the config specifically for Storybook.

sirreal commented on packages/components/src/product-icon/index.stories.jsx:
```diff
@@ -0,0 +1,23 @@
+	return (
+		<>
+			{ supportedSlugs.map( slug => (
+				<div className="index.stories__icon-tile">
```
Why does this class include a literal `.`? That seems unorthodox 😮

jameslnewell replied:
The linting enforces it, I think based on the file name - `index.stories.jsx`

jsnajdr replied:
In this case, the rule also allows `product-icon` (the surrounding folder name), which is a more wise choice of a CSS class name.

I filed #38764 to prevent ESLint from suggesting the less reasonable `index.stories` name.

jameslnewell replied:
Are we adhering to BEM? If we are I'd argue that the `index.stories` block and elements aren't part of the `product-icon` block, particularly because `index.stories__icon` _contains_ a `product-icon` block.

While I found the rule and the dot syntax unexpected too, it makes sense to me that the storybook specific styling is seperate from the actual component styling.

sirreal commented on packages/components/src/product-icon/index.stories.scss:
```diff
@@ -0,0 +1,21 @@
+
+.index\.stories {
```
🙈

sirreal commented on client/assets/stylesheets/shared/mixins/_mixins.scss:
```diff
@@ -1,8 +1,8 @@
+@import '~@automattic/components/src/styles/_breakpoints.scss';
 @import 'calc';
-@import 'clear-fix';
+@import '~@automattic/components/src/styles/_clear-fix.scss';
```
I'd like to know whether you have longer term thoughts on this change? It seems that we have moved some components to the package arbitrarily (we moved them as they were needed and when they were considered "leaf" components), and now we are moving some arbitrary bits of SASS as well.

Would we expect all of the "framework"-type SASS to move to the components package?

---

I don't really like importing from an arbitrary `src` path, that seems a bit like accessing private package internals. I wouldn't want e.g. the components to be used from `src` 😕 

If we have SASS that is public in this package, I'd suggest it go in a directory that makes that clear. I also wonder if there would be benefit from categorization under this directory like we have in Calypso, but I don't feel strongly about that:

```sass
// "Public" directory
@import '~@automattic/components/sass/_breakpoints.scss';
// Categorized
@import '~@automattic/components/sass/mixins/_breakpoints.scss';
```

jameslnewell replied:
> I'd like to know whether you have longer term thoughts on this change? 

In general my long term hope is that the use of these utilities outside the package itself is greatly reduced as we offer more granular components that can be composed therefore, reducing the need for additional styling and these utilities. 

That way we might be able to stop exporting some of the utilities.

The remaining utilities could be grouped and exposed in their own packages e.g.

```scss
@import '~@a8c/breakpoint`;

@debug $a8c-breakpoints;

@include a8c-breakpoint('>large') {
  // ...
}
```

> Would we expect all of the "framework"-type SASS to move to the components package?

There's a good chance that it might all move depending on what parts are deemed reusable! 

I'm keen to incrementally pull across and refactor what we need rather than moving everything across up front and exposing it - that makes it harder for Calypso to iterate, adds more code to be maintained in `@automattic/components` and increases the likelyhood of what is exposed requiring breaking changes.

I'm keen to only move what's necessary across leaving Calypso free to change and iterate without impacting others.

> If we have SASS that is public in this package, I'd suggest it go in a directory that makes that clear. 

I like it 👍 

> I also wonder if there would be benefit from categorization under this directory like we have in Calypso, but I don't feel strongly about that

My personal preference is definitely to group by concern (e.g. typography, breakpoints) rather than type (e.g. variable, mixin, function) because concerns are usually coupled and it saves me looking across multiple files and forming a mental picture of how code across multiple files fits together... but if there's a prior pattern I'm OK with towing the line!

p-jackson replied:
I actually like this idea of only moving the modules that are needed for now. Similar to what I said about the changes in `_typogrpahy.scss`, it could be that we never end up sharing those files with anything outside calpyso.
I think we can develop stronger ideas about what belongs in and out of `@automattic/components` once we've been using it for longer.

However I do agree that since this is now part of the package's public API that the files shouldn't be getting imported from `src`.

simison replied:
> it could be that we never end up sharing those files with anything outside calpyso.

I could see us use shared Typography sass to make things consistent across Calypso and WP Admin for Atomic sites.

p-jackson replied:
> I could see us use shared Typography sass to make things consistent across Calypso and WP Admin for Atomic sites.

That's true and would be pretty awesome. By the time we are ready for that we could be including it from somewhere like `@8c/typography`. Or maybe it'll be a different solution. I'm just biased towards deciding that stuff as late as possible.

jameslnewell replied:
BTW a `__experimentalText` component just merged yesterday. Unfortunately the `text()` mixin  (CSS-in-JS)isn't exposed yet, but the component should be able to support a bunch of use cases by itself.

https://github.com/WordPress/gutenberg/pull/18495

sirreal commented on client/assets/stylesheets/shared/_typography.scss:
```diff
@@ -7,11 +8,7 @@ $code: Monaco, Consolas, 'Andale Mono', 'DejaVu Sans Mono', $monospace;
 
-$sans: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen-Sans', 'Ubuntu', 'Cantarell',
-	'Helvetica Neue', sans-serif;
-$sans-rtl: Tahoma, $sans;
```
Extracting `sans` and leaving the rest of these general font declarations seems strange. Do you think `$monospace`, `$code`, `$serif`, and `$serif-fallback` should move as well?

jameslnewell replied:
I'm keen to move them when `@automattic/components` needs them. I'd prefer to keep Calypso specific code out of `@automattic/components` in order to keep what's exposed by the package minimal, leaving the code free to change and iterate as necessary in Calypso.

p-jackson replied:
I also like the idea of keeping `$monospace` etc. where is until it's actually shared by something. It could be that those fonts are never get used by a shared component, and that's fine.

Commit:
fix variable and mixin naming

Commit:
fix line break

p-jackson reviewed (commented):
Like I said above, I prefer the approach you've taken of only moving the files that are necessary in this PR. But I can also understand the rationale of planning out how we want these packages to work now and in the future, so happy for that to go either way.

I also did a quick visual look at calypso/devdocs/storybook. Too much to test manually, but it doesn't look completely broken :)

I did notice that `<Dialog>` in storybook doesn't look like what I'm used to in Calypso. Is that because there's still some style that hasn't been pulled into the package, or because `<Dialog>` is supposed to look a little unstyled?

jameslnewell commented on packages/components/src/button/README.md:
````diff
@@ -7,6 +7,7 @@ Buttons express what action will occur when the user clicks or taps it. Buttons
 ```jsx
 import Gridicon from 'components/gridicons';
 import { Button } from '@automattic/components';
+import '@automattic/calypso-color-schemes/dist/calypso-color-schemes.css';
````
Importing the color schemes in `Button` so the consumer doesn't have to do it sounds like a great idea which I was so close to start coding!

I just realised unfortunately that since the `@import` defines a bunch of global CSS variables we only want to `@import` it once in the entire application. If we `@import` it in `Button`, the consumer cannot import it anywhere. That gets more complicated when you consider that other components require that `@import` and the possibility for multiple versions of the `@automattic/calypso-color-schemes` package to be used within the same app. 

<details>
<summary>🤔</summary>
<br/>
:trollface: A CSS-in-JS theme provider would nicely solve these pain points
</details>

Can you think of a better way to document in `@automattic/components` that `@automattic/calypso-color-schemes/dist/calypso-color-schemes.css` needs to manually be imported once per application?

p-jackson replied:
Ah theming 😭

I think package level documentation is needed in general (not this PR though) since we require some things from the client e.g. supported React versions, a sass-loader is required, etc. And I think just add to that list that clients must include this magic file once in their app.

But alas I think you're right, the constants just need to be defined once per app. I was hoping that tree-shaking magic would be able to help us with this.

jameslnewell replied:
I've removed the import from the `Button` readme and added it in a `@automattic/components` readme.

simison replied:
>To solve this, can you think of a better way to document in @automattic/components that @automattic/calypso-color-schemes/dist/calypso-color-schemes.css needs to manually be imported once per application?

`@automattic/calypso-color-schemes` is problematic for re-use because it contains very WordPress.com specific CSS vars; so it's not actually "framework package" like the name suggests. Things like "sidebar background-color" etc.

Those then end up in each bundle where we need a mere button, e.g. in Jetpack.

`@automattic/color-studio` seems better that way but still at on of CSS variables that end up not being used.

jameslnewell replied:
👍 Yep, I agree, long term the variable naming isn't amazing, in some places its overly generic and in others its overly specific but its currently a dependency of the components.

Longer term I'd like to see a minimal set of variables that can be customised by the consumer - similar to Material's idea of surfaces. @davewhitley and I were [experimenting](https://salad-ui-playground.now.sh/plugins/featured) with these [ideas](https://salad-ui.now.sh/foundations/color/) (click the icon in the bottom right to toggle the variables).

For now, asking the consumer to import the variables once universally in the app feels like an acceptable workaround until we get around to deciding on and refactoring those things.

jameslnewell commented:
> I also did a quick visual look at calypso/devdocs/storybook. Too much to test manually, but it doesn't look completely broken :)

🙏 Thanks @p-jackson!

> I did notice that <Dialog> in storybook doesn't look like what I'm used to in Calypso. Is that because there's still some style that hasn't been pulled into the package, or because <Dialog> is supposed to look a little unstyled?

I saw there was an [`EmbedDialog` in `/devdocs`](https://wpcalypso.wordpress.com/devdocs/design/embed-dialog) but I don't have anything to reference for a `Dialog`.

I couldn't find any styles applicable to `.dialog` in the Calypso reset/base styles so I don't think its missing anything... but the `Dialog` component does have a `baseClassName` class which allows the consumer to specify their own classname prefix to change alltheclassnames. I've got no idea what those could be at the moment.

jameslnewell commented:
I tried moving `src/styles/*` to `sass/*` but the generated `dist/[esm|cjs]` directory structure means that the files are one level higher and the relative path needs to be `../../../sass/typography.scss` instead of `../../sass/typography.scss`. 

Any ideas on how to support that without requiring consuming developers to update their build pipelines? 

/cc @sirreal @p-jackson 

<details>
  <summary>🤔</summary>
  <br/>
   :trollface: moving the mixins into CSS-in-JS would avoid these problems 
</details>

Commit:
added a readme for @automattic/components

sirreal commented:
> moving them into CSS-in-JS would be an easy way to avoid this problem
> A CSS-in-JS theme provider would nicely solve these pain points

I've been feeling these pain points for a while and am largely convinced by css-in-JS. I think we're in a position to seriously consider that alternative rather than continuing to struggle with this when the tradeoffs clearly point to using css-in-JS. Have you considered what that approach would look like here?

There is precedent for using [emotion in Calypso packages](https://github.com/Automattic/wp-calypso/blob/ac9027fcd71b51997067b1f4b2c5bce603deb5ad/packages/composite-checkout/src/components/checkout.js#L232-L269) already, and the current guidelines seem very open to exploring this space when it makes sense: pbjpUB-9x-p2

cc: @griffbrad

p-jackson commented:
> Any ideas on how to support that without requiring consuming developers to update their build pipelines?

I'm think maybe you could create scss files in `components/sass`, and then have them `@import` the styles from `components/src/styles`.

The internal style files can be split, rearranged, maybe even re-exported from something like `@a8c/breakpoints` without having to break backwards compatibility.

Something like this:
https://gist.github.com/p-jackson/e3efa29bc097f29a05c9e2b53625e83c

@jameslnewell

diegohaz commented:
Is it possible to import `@automattic/calypso-color-schemes` without requiring users to specify `path/to/css/file.css`?

I also agree that CSS-in-JS would solve many of the problems that are being discussed here.

jameslnewell commented:
> Have you considered what that approach would look like here?

@sirreal I haven't thought about in depth but simply I think it'd look like:

- Create a `theme` for Calypso to be used with `emotion`'s `ThemeProvider` - potentially a good time to rethink variable naming/usage (as discussed above)
- Rewrite the necessary mixins as JavaScript functions - I probably wouldn't update Calypso to use these (as I have in this PR), instead when a team migrates a section of Calypso to CSS-in-JS they can use whatever mixins the `@automattic/components` (or `@a8c/*` as discussed previously in this PR) might expose.
- Rewrite the components in CSS-in-JS themselves - need to decide if we want to keep the class names around for backwards compatibility

jameslnewell commented:
Good idea @p-jackson will have a crack at it!

Commit:
remove file path for color-schemes import

jameslnewell commented:
> Is it possible to import @automattic/calypso-color-schemes without requiring users to specify path/to/css/file.css?

@diegohaz seems like it can thanks to the `main`. Updated.

Commit:
expose mixins via sass folder

Commit:
update mixin naming in merges

p-jackson reviewed (commented):
Everything's looking good to me now in the code @jameslnewell 💪

I'm a bit nervous about the test failures. Given that this PR is too big to test manually we're really leaning on the e2e tests to reassure us. I even restarted them a couple times but, they still failed.

Test failures are in Gutenberg. When I do some trivial page edits nothing stands out to me as completely broken, but I think we should figure out what's going on.

p-jackson commented on packages/components/README.md:
````diff
@@ -2,6 +2,31 @@
+
+```jsx
+// import the color variables - you only need to do this once in your entire application
+import '@automattic/calypso-color-schemes/dist/calypso-color-schemes.css';
````
```suggestion
import '@automattic/calypso-color-schemes';
```

p-jackson reviewed (approved):
I restarted the tests enough times to get them to pass! 🤣
I think the issue was that there was something wrong with the store sandbox used by tests.

jameslnewell commented:
Conflicts gone and tests passing 🎉🎉🎉

Do we want to merge or close this PR given that this PR moves and renames SASS mixins from the `calypso` app into the `@automattic/components` package? When these components get migrated to CSS-in-JS we'll probably move those SASS mixins back to the `calypso` app and duplicate them in JS for `@automattic/components`.

Are we better off doing that straight away, before we expose the mixins and people come to rely on the exports?

/cc @sirreal

sirreal commented:
> Do we want to merge or close this PR given that this PR moves and renames SASS mixins from the `calypso` app into the `@automattic/components` package?When these components get migrated to CSS-in-JS we'll probably move those SASS mixins back to the `calypso` app and duplicate them in JS for `@automattic/components`.

Would it be an option to keep the added SASS necessary in `@automattic/components`, but revert the changes that moves/renames them out of Calypso? This seems desirable if moving them back is an imminent next step.

I don't see a big risk in merging this (potentially with the preceding suggestion) and publishing a version if folks want to use it. I'd like us to get to a CSS-in-JS solution soon because that all of our experience trying to share components so far is pointing to that being the best solution.

> Are we better off doing that straight away, before we expose any SASS mixins and people come to rely on them from the package (which will require a plan to migrate in the future)?

I'd be fine with releasing a version with SASS then removing it in a subsequent version. Version numbes are cheap, even major ones 🙂

Commit:
reverted moving and renaming mixins from calypso

jameslnewell commented:
I've reverted the renaming and moving of variables and mixins in Calypso. As a result the necessary variables and mixins are duplicated internally within the `@automattic/components` package. There's no need to export the mixins anymore so I've removed the `sass` directory.

I'll start switching the components to CSS-in-JS in future PRs where I'll migrate the duplicated SASS mixins to CSS-in-JS.

Commit:
reset package-lock to master

Commit:
updated package-lock
